pcsx2/Recording/InputRecordingFile.cpp

```diff
@@ -24,119 +24,158 @@
-			StateCopy_SaveToFile(path + "_SaveState.p2s");
-		}
-		else
-		{
-			sApp.SysExecute();
-		}
-	}
-	return true;
+void InputRecordingFileHeader::SetEmulatorVersion()
+{
+	wxString emuVersion = wxString::Format("%s-%d.%d.%d", pxGetAppName().c_str(), PCSX2_VersionHi, PCSX2_VersionMid, PCSX2_VersionLo);
+	int max = ArraySize(emu) - 1;
+	strncpy(emu, emuVersion.c_str(), max);
+	emu[max] = 0;
+}
+
+void InputRecordingFileHeader::SetGameName(wxString _gameName)
+{
+	int max = ArraySize(gameName) - 1;
+	strncpy(gameName, _gameName.c_str(), max);
+	gameName[max] = 0;
 }
 
-// Gracefully close the current recording file
 bool InputRecordingFile::Close()
 {
 	if (recordingFile == NULL)
 	{
 		return false;
 	}
 	WriteHeader();
-	WriteSaveState();
+	writeSaveState();
 	fclose(recordingFile);
 	recordingFile = NULL;
 	filename = "";
 	return true;
 }
 
-// Write savestate flag to file
-bool InputRecordingFile::WriteSaveState()
+const wxString& InputRecordingFile::GetFilename()
+{
+	return filename;
+}
+
+InputRecordingFileHeader& InputRecordingFile::GetHeader()
+{
+	return header;
+}
+
+unsigned long& InputRecordingFile::GetTotalFrames()
+{
+	return totalFrames;
+}
+
+unsigned long& InputRecordingFile::GetUndoCount()
+{
+	return undoCount;
+}
+
+bool InputRecordingFile::FromSaveState()
 {
+	return savestate.fromSavestate;
+}
+
+void InputRecordingFile::IncrementUndoCount()
+{
+	undoCount++;
 	if (recordingFile == NULL)
 	{
-		return false;
+		return;
 	}
+	fseek(recordingFile, seekpointUndoCount, SEEK_SET);
+	fwrite(&undoCount, 4, 1, recordingFile);
+}
 
-	fseek(recordingFile, RecordingSeekpointSaveState, SEEK_SET);
-	if (fwrite(&savestate.fromSavestate, sizeof(bool), 1, recordingFile) != 1)
+bool InputRecordingFile::open(const wxString path, bool newRecording)
+{
+	if (newRecording)
+	{
+		if ((recordingFile = wxFopen(path, L"wb+")) != nullptr)
+		{
+			filename = path;
+			totalFrames = 0;
+			undoCount = 0;
+			header.Init();
+			return true;
+		}
+	}
+	else if ((recordingFile = wxFopen(path, L"rb+")) != nullptr)
 	{
+		if (verifyRecordingFileHeader())
+		{
+			filename = path;
+			return true;
+		}
+		Close();
+		recordingConLog(wxString::Format("[REC]: Input recording file header is invalid\n"));
 		return false;
 	}
-
-	return true;
+	recordingConLog(wxString::Format("[REC]: Input recording file opening failed. Error - %s\n", strerror(errno)));
+	return false;
 }
 
-// Write controller input buffer to file (per frame)
-bool InputRecordingFile::WriteKeyBuf(const uint& frame, const uint port, const uint bufIndex, const u8& buf)
+bool InputRecordingFile::OpenNew(const wxString path, bool fromSavestate)
 {
-	if (recordingFile == NULL)
+	if (fromSavestate)
 	{
-		return false;
+		if (CoreThread.IsOpen())
+		{
+			if (open(path, true))
+			{
+				savestate.fromSavestate = true;
+				if (wxFileExists(path + "_SaveState.p2s"))
+				{
+					wxString backupPath = path + "_SaveState.p2s.bak";
+					while (wxFileExists(backupPath))
+						backupPath += ".bak";
```

Comment:
I think this should be an `if` instead of a `while`

Reply:
The problem with using an if here is that if there is already a backup in place, the existing savestate would then not get backed up due to how wxCopyFile works. By using a while, it ensures that no matter what, a backup of the existing savestate is saved by adding an extra ".bak" tag.
If the behavior was instead that the existing backup was overridden, then I would be using "if," but that's sadly not the case.

Reply:
Oooooo and there's a boolean parameter I can use for overwrite. Poggers. Turning that to an if.

Reply:
ah i hadn't considered that, but nice to know there's an even better workaround.  good find